### ISSUE TITLE
Perf issues opening thousands of documents

### ISSUE THREAD

**Issue by saurabh-sharma** [Microsoft/vscode#35996]
- VSCode Version: Code 1.17.0 (be377c0faf7574a59f84940f593a6849f12e4de7, 2017-10-04T23:33:45.303Z)
- OS Version: Linux x64 4.10.0-37-generic
- Extensions: the listing length exceeds browsers' URL characters limit
---

VSCode consuming high amount of CPU & memory on startup and becomes unresponsive.
I get "Window is not responding" message

Here's the CPU usage snapshot
![screenshot from 2017-10-10 23-23-21](https://user-images.githubusercontent.com/3684584/31402230-08aaf1a6-ae13-11e7-8b2f-7476c36fd281.png)

Steps to Reproduce:
1. Open VSCode
2. Critical unresponsiveness is observed when vscode opens with an existing project with some files open.

Reproduces without extensions: No

**Comment by egamma** [comment]
@saurabh-sharma what kind of project (language, size) is this?

Since you mention that this cannot be reproduced without extensions, can you please help us to narrow down which extension is causing this: uninstall an extension, restart, observe whether problem persists. I understand that this painful since you have many extension installed. I would start with the extension that provides language services.

**Comment by saurabh-sharma** [comment]
@egamma I am using React over my project.
The issue was being caused by [IntelliSense for CSS class names](https://marketplace.visualstudio.com/items?itemName=Zignd.html-css-class-completion)

On removing the extension, vscode worked fine.

Although the same is working fine with the version prior to VSCode Version: Code 1.17.0 (be377c0, 2017-10-04T23:33:45.303Z)

**Comment by egamma** [comment]
@saurabh-sharma thanks for following up. The fact that you say the extension wasn't a performance issue before 1.17 is a valuable hint that would help us to track the root cause of this issue. 

- is this an project that you can share and that we can profile? If not can you create a CPU profile for us (steps below).
- in your settings do you configure files.exclude or search.exclude?
- in your .gitignore file do you exclude node_modules?

If the workspace is not sharable can you do a CPU profile when the extension is installed. These are the Hit F1 > Toggle Developer Tools. In the overflow menu (see below) select 'More Tools > JavaScript Profiler'. In there select start, let it profile for 30 to 60 seconds, stop it, save it to a file and attach the file here. Since the problem happens on start-up reload the window after you started the profiler using F1>Reload Window.

<img width="380" alt="screen shot 2017-09-28 at 09 44 31" src="https://user-images.githubusercontent.com/1794099/30954796-d1be9e30-a431-11e7-959e-495d234c37c6.png">

**Comment by egamma** [comment]
The [extension](https://marketplace.visualstudio.com/items?itemName=Zignd.html-css-class-completion) that triggers the CPU load is using: `vscode.workspace.findFiles(`**/*.{${languages}}`, '')` and in 1.17 we have switched to use ripgrep. One hypothesis is that the switch to ripgrep caused that we are no longer excluding node modules.
In 1.17 we have switched to use ripgrep.

**Comment by egamma** [comment]
The PR for the extension has been submitted https://github.com/zignd/HTML-CSS-Class-Completion/pull/89

**Comment by egamma** [comment]
To reproduce the issue in VS Code.

- Install the HTML Hint, Markdown Lint extensions (they will report issues on node modules)
- In VS Code 1.16, install the CSS Intellisense extension
- Git clone a repository with node modules with some markdown/html files. I've used https://github.com/Microsoft/vscode-npm-scripts.
- Open this workspace in VS Code
Observe that the Problems View is empty.

Now do the same steps in VS Code 1.17. The Problems View will contain many errors, which shows that workspace.findFiles also returns files inside the node_modules folder. These files are then opened by the extension which triggers the linters.

**Comment by chrmarti** [comment]
I see the problem reported in 1.16.1 too:
![image](https://user-images.githubusercontent.com/9205389/31502353-a9a4febc-af21-11e7-82d9-c816d5c3d43e.png)

**Comment by roblourens** [comment]
I think in https://github.com/Microsoft/vscode/issues/34434 and https://github.com/Microsoft/vscode/issues/34487 we discovered that node_modules and .git/ was never excluded, so I don't think that aspect of it is new.

**Comment by chrmarti** [comment]
`findFiles()` does apply `files.exclude` but not `search.exclude` which lists `node_modules` by default. My screenshot above the code indicate that `node_modules` wasn't excluded in 1.16.x.

@saurabh-sharma The CSS Intellisense extension just published a new version with the suggested fix. Could you check if you still see the problem with that newer version?

**Comment by egamma** [comment]
@chrmarti hmm, below is a screen shot from my setup and it shows that node_modules was excluded
![image](https://user-images.githubusercontent.com/172399/31506460-c89e1942-af77-11e7-9bd2-f92bc318dee0.png)

**Comment by chrmarti** [comment]
@egamma Did the extension auto-update in your install? Your change is already published in a new version. Do you have `files.exclude` set to something else than the default?

@jrieken When I open one of my large workspaces the 'Intellisense for CSS class names' extension quickly gets both 1.16 and 1.17 locked up. Spending a lot of time comparing URIs in: https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/api/electron-browser/mainThreadDocumentsAndEditors.ts#L107

Still looking for the regression.

**Comment by jrieken** [comment]
I'd say that's the effect of opening many, many documents. On every change we need to sync the documents and editor over. @chrmarti Do you have a real CPU profile that you can share?

**Comment by chrmarti** [comment]
@jrieken I do and I think you're right. I wonder if computing the diff with a set/map instead of sorted arrays would be faster. I've copied the profiles I currently have to //mt1/eng/35996, the one I'm currently looking at is the `-master` which is from unminified source.

@sandy081 Another hot spot in the profile is `ConfigurationModel.override()` that gets triggered from `FileService.configuredAutoGuessEncoding()`. That I haven't seen in the 1.16.1 profile, but I need to double-check.

I don't think we can explain the reported regression yet. 1.16 performs just as badly with that many opened documents as 1.17.

**Comment by chrmarti** [comment]
![image](https://user-images.githubusercontent.com/9205389/31510036-5459adc4-af38-11e7-8024-52233d1ddc6a.png)

**Comment by diminutivesloop** [comment]
I'm seeing a similar issue, but it manifests even when I have no open projects and all extensions are disabled.
![vscode startup cpu usage](https://user-images.githubusercontent.com/667716/31511450-c51e6c3a-af55-11e7-99d9-16982c40dc67.PNG)

**Comment by chrmarti** [comment]
@diminutivesloop That looks like a different problem, could you open an new issue (Help > Report Issues)? Thanks.

**Comment by chrmarti** [comment]
@sandy081 `ConfigurationModel.override()` could clone the contents only when the first override that needs to be merged is found. For `FileService.configuredAutoGuessEncoding()` this could be optimized further because that only needs a single value, no need to compute the full config first.

The document sorting is the bigger issue though.

**Comment by jrieken** [comment]
fyi - @sandy081 is doing a bigger refactoring such that events are more fine grained and that less object clones happen

**Comment by jrieken** [comment]
> @jrieken I do and I think you're right. I wonder if computing the diff with a set/map instead of sorted arrays would be faster.

In theory the sorting takes O(n log n) and the delta should take O(n) tho I do see that the `sortedDiff`-function creates tons of arrays. @joaomoreno sure that you made a good change there?

**Comment by chrmarti** [comment]
If you don't sort you can do it in O(n). I've posted prototype here: #36162. With that the delta computation still is the most expensive part in the profiler. But I'm also using a huge workspace and the extension is opening all its html and css documents in parallel.

**Comment by egamma** [comment]
Rather then trying to change the VS Code for the recovery build, we have investigated into improving the indexing performance of the extension and  submitted another PR https://github.com/zignd/HTML-CSS-Class-Completion/pull/90.

I'll therefore remove this from the 1.17.2 candidate list.

**Comment by VsTB** [comment]
INGMASTER© SERVER 
IMS™ Home free network

vstb@gialgiay.ru
+79287273737

GIALGIAY MOHK INGUSHETIA MALGOBEK
INGMASTER© Development of web sites VGroup
Classify the system administrator

**Comment by chrmarti** [comment]
Over to @jrieken and @joaomoreno for the relatively expensive delta computation.

**Comment by saurabh-sharma** [comment]
@chrmarti After installing the CSS Intellisense extension update, vscode is running fine as it should.

**Comment by sandy081** [comment]
@chrmarti

> @sandy081 Another hot spot in the profile is ConfigurationModel.override() that gets triggered from FileService.configuredAutoGuessEncoding(). That I haven't seen in the 1.16.1 profile, but I need to double-check.

Encoding setting is made lang specific setting in 1.17 and hence the setting is being read from overrides. Any idea what proportion of the regression time, reading settings is being taking?

> @sandy081 ConfigurationModel.override() could clone the contents only when the first override that needs to be merged is found. For FileService.configuredAutoGuessEncoding() this could be optimized further because that only needs a single value, no need to compute the full config first.

Yes, we can get the override contents smartly. 

I am doing a big refactoring in configuration area where I am also fixing such issues.

**Comment by chrmarti** [comment]
@sandy081 Not sure about the overall proportion, but the model sorting clearly dominated.

**Comment by Esteban-Rocha** [comment]
+1 here

Same issue with:

Version 1.17.1
Commit 1e9d36539b0ae51ac09b9d4673ebea4e447e5353
Date 2017-10-10T14:21:18.489Z
Shell 1.7.7
Renderer 58.0.3029.110
Node 7.9.0
Architecture x64

System:
Arch linux, 64x
iCore7 - 16GB RAM

After activating this extension CPU increases 40% to 50%, RAM also suffers a 500MB to 1GB increase.

I'm a WP Dev I do not use npm a lot, In many projects it doesn't even exist, still I get this result, so you may want to have that into consideration about the node_modules directory issue.

**Comment by joaomoreno** [comment]
> In theory the sorting takes O(n log n) and the delta should take O(n) tho I do see that the sortedDiff-function creates tons of arrays. @joaomoreno sure that you made a good change there?

@jrieken Yeah. Is the array allocation something that is making an impact?

**Comment by jrieken** [comment]
Was just blaming you @joaomoreno because you were the last to touch any code in that area. And yes do believe creating too many array isn't good. 

I did the following, write an extension that find all typescript files and opens them. Run that on our vscode workspace with ~4400 files. 

```ts
 vscode.commands.registerCommand('extension.sayHello', () => {
        const t1 = Date.now();
        let total = 0;
        vscode.workspace.findFiles('{**/*.ts}').then(resources => {
            console.log('total', resources.length)
            const p = resources.map((resource, i) => {
                return vscode.workspace.openTextDocument(resource).then(doc => {
                    total += doc.getText().length;
                })
            });

            return Promise.all(p).then(() => console.log('DONE -> ms' + (Date.now() - t1)));
        });
    })
```

The existing solution is the base, mostly because of `compareModels` which I blame on quick sort. Maybe those arrays are nearly sorted and quick sort runs in O(n2)... I have then replaces that with our super-naive merge sort implementation and things a getting better (40sec vs 46sec). The [delta-based approach](https://github.com/Microsoft/vscode/pull/36162/files) is the worst but that is very likely because the new `delta` method end in optimisation hell. 

The archive contains all profiles, the name hints the approach and the runtime: [Archive.zip](https://github.com/Microsoft/vscode/files/1394887/Archive.zip)

@joaomoreno  is off the hook, others will get on it as there are more hot code paths in this...

**Comment by jrieken** [comment]
I did improve the editor/document sync logic to be fit for the document add case... This are still slow, very slow. @bpasero should get involved as a whole minute is spend in some event processing. 

<img width="1111" alt="screen shot 2017-10-27 at 18 27 25" src="https://user-images.githubusercontent.com/1794099/32114808-bfb057ae-bb44-11e7-8111-fd3036536114.png">
[after.cpuprofile.txt](https://github.com/Microsoft/vscode/files/1422736/after.cpuprofile.txt)

**Comment by bpasero** [comment]
@jrieken tried with your steps and those `textFileEditorModel` methods are not showing up in the top consuming methods for me, can you reproduce reliably and if so is there more to it than running https://github.com/Microsoft/vscode/issues/35996#issuecomment-337604103 on the vscode workspace?

My trace: 
[CPU-20171030T120534.cpuprofile.txt](https://github.com/Microsoft/vscode/files/1426928/CPU-20171030T120534.cpuprofile.txt)

Looking at your trace, the top consuming method seems to be this [`onStateChange`](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/services/textfile/common/textFileEditorModel.ts#L127) handler which should be a no-op unless files are being reverted.

**Comment by jrieken** [comment]
Yes, always reproduces. @bpasero Did you really run this on our workspace with thousands of documents (4414 to be precise)? Your longest single method runs for just a 1sec, I have runtime in the area of 10-30seconds.

**Comment by bpasero** [comment]
@jrieken tried on vscode workspace and I am seeing ~1700 files returned but I can try with more files.

**Comment by bpasero** [comment]
@jrieken seems better to me after https://github.com/Microsoft/vscode/commit/40b58afdcfe747aa851c3b1dc079e7f037ef4e89 can you check?

**Comment by jrieken** [comment]
Yeah, seems a lot better now, mostly garbage collecting (we should memory profile this) and object-clone (/cc @sandy081 on that)

[after.cpuprofile.txt](https://github.com/Microsoft/vscode/files/1427833/after.cpuprofile.txt)
<img width="1050" alt="screen shot 2017-10-30 at 16 51 29" src="https://user-images.githubusercontent.com/1794099/32180585-aa645b5c-bd92-11e7-8141-884d751d653d.png">

**Comment by bpasero** [comment]
I also noticed the cloning appearing to the top.

**Comment by sandy081** [comment]
@bpasero @jrieken 

Configuration is always cloned when is being asked, otherwise there could be a chance of changing the value in-memory which is error-prone. May I know where are the majority of the calls reading configuration are coming from ?

**Comment by bpasero** [comment]
Looks like some calls are originating from file service finding out the configured encoding for each resource that gets loaded ([here](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/services/files/node/fileService.ts#L651)).

**Comment by jrieken** [comment]
> Configuration is always cloned when is being asked, otherwise there could be a chance of changing the value in-memory which is error-prone.

Why not call object.freeze/deepfreeze once when creating those objects?

**Comment by sandy081** [comment]
@jrieken I have to do deepFreeze, but did not find a native deepFreeze method. If we do deepFreeze ourselves, won't be similar to deepClone?

**Comment by jrieken** [comment]
We used to have `deepFreeze` in objects.ts but we might have removed (unused). The idea is to walk the object recursively and freeze it in place (`Object.freeze` modifies the argument)

> If we do deepFreeze ourselves, won't be similar to deepClone?

Well, you only do it once, when you compute/write the data, but not everytime you are asked for configuration (reading).

**Comment by sandy081** [comment]
Configuration service was made faster by not cloning always instead of freezing once and surfacing frozen object.
